js/packages/openinference-vercel-ai-sdk-span-processor/src/OpenInferenceSpanProcessor.ts

```diff
@@ -0,0 +1,41 @@
+import { ReadableSpan, SpanProcessor } from "@opentelemetry/sdk-trace-base";
+import {
+  getOIModelNameAttribute,
+  getOISpanKindFromAttributes,
+  hasAIAttributes,
+} from "./utils";
+import { Span } from "@opentelemetry/api";
+import { SemanticConventions } from "@arizeai/openinference-semantic-conventions";
+
+export class OpenInferenceSpanProcessor implements SpanProcessor {
+  async forceFlush() {
+    // No-op
+  }
+
+  onStart(_: Span): void {
+    // No-op
+  }
+
+  async shutdown(): Promise<void> {
+    // No-op
+  }
+  onEnd(span: ReadableSpan): void {
+    const initialAttributes = span.attributes;
+    if (!hasAIAttributes(initialAttributes)) {
+      return;
+    }
+
+    const spanKind = getOISpanKindFromAttributes(initialAttributes);
+
+    if (spanKind == null) {
+      return;
+    }
+    // @ts-expect-error - This is a read-only span and thus has no setter for attributes
+    // Manually patch attributes here
+    span.attributes = {
+      ...span.attributes,
+      ...getOIModelNameAttribute(initialAttributes),
+      [SemanticConventions.OPENINFERENCE_SPAN_KIND]: spanKind,
+    };
```

Comment:
@mikeldking onEnd exposes a ReadableSpan which has this type structure. It does not have setters for attributes and the attributes are "typescript level" readonly

```ts
export interface ReadableSpan {
    readonly name: string;
    readonly kind: SpanKind;
    readonly spanContext: () => SpanContext;
    readonly parentSpanId?: string;
    readonly startTime: HrTime;
    readonly endTime: HrTime;
    readonly status: SpanStatus;
    readonly attributes: SpanAttributes;
    readonly links: Link[];
    readonly events: TimedEvent[];
    readonly duration: HrTime;
    readonly ended: boolean;
    readonly resource: IResource;
    readonly instrumentationLibrary: InstrumentationLibrary;
    readonly droppedAttributesCount: number;
    readonly droppedEventsCount: number;
    readonly droppedLinksCount: number;
}
```